File: lint-arwaky/src-python/taxonomy/git_hook_error.py
"""git_hook_error — Git hook management domain error types."""

from pydantic import BaseModel, ConfigDict

from .error_code_vo import ErrorCode
from .error_value_vo import (
    ErrorMessage,
    Cause,
)
from .file_path_vo import FilePath


class GitHookError(BaseModel):
    """General failure during Git hook management."""

    model_config = ConfigDict(frozen=True)

    path: FilePath | None = None
    message: ErrorMessage
    error_code: ErrorCode | None = None
    cause: Cause | None = None

    def __str__(self):
        target = f" on {self.path}" if self.path else ""
        code = f" [{self.error_code}]" if self.error_code else ""
        return f"Git Hook Error{target}{code}: {self.message}"

File: lint-arwaky/src-rust/taxonomy/git_hook_error.rs
use serde::{Serialize, Deserialize};
use super::*;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct GitHookError {
    #[serde(default)]
    pub path: Option<FilePath>,
    pub message: ErrorMessage,
    #[serde(default)]
    pub error_code: Option<ErrorCode>,
    #[serde(default)]
    pub cause: Option<Cause>,
}

impl GitHookError {
    pub fn new(message: ErrorMessage) -> Self {
        Self { path: None, message, error_code: None, cause: None }
    }
}

impl std::fmt::Display for GitHookError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let target = self.path.as_ref().map(|p| format!(" on {}", p)).unwrap_or_default();
        let code = self.error_code.as_ref().map(|c| format!(" [{}]", c)).unwrap_or_default();
        write!(f, "Git Hook Error{}{}: {}", target, code, self.message)
    }
}
